README.md
# node-es

This is a Node.js module for the [elasticsearch](http://www.elasticsearch.org/) REST API.

NOTE: `node-es` `v0.6` and newer work with ElasticSearch 5 and up.  For older versions of ElasticSearch, prior versions of `node-es` should be used.

[![Build Status](https://travis-ci.org/ncb000gt/node-es.png)](https://travis-ci.org/ncb000gt/node-es) [![Coverage Status](https://coveralls.io/repos/ncb000gt/node-es/badge.png)](https://coveralls.io/r/ncb000gt/node-es)

## Install

```Javascript
npm install es
```

### Elasticsearch Version Compatibility

When working with Elasticsearch v7.x and up, use `v0.8.x` (latest `npm install es`). When using previous versions of Elasticsearch, please use `v0.7.4` of this module (`npm install es@v0.7.4`).

## Usage

```Javascript
var
  elasticsearch = require('es'),
  config = {
    _index : 'kittehs'
  },
  es = elasticsearch(config);

es.search({
    query : {
      field : {
        animal : 'kitteh'
      }
    }
  }, function (err, data) {
    // work with data here
    // response data is according to ElasticSearch spec
  });
```

## API

Unless otherwise stated, all callback signatures are `function (err, data)`, with `data` being the parsed JSON response from elasticsearch.

#### createClient

Calling `elasticsearch.createClient(config)` is the same as `elasticsearch(config)`.

```Javascript
var
  elasticsearch = require('es'),
  es = elasticsearch.createClient(config);
```

##### config._index

When initializing the library, you may choose to specify an index to work with at the start to save from having to supply this information in the options for each operation request:

```Javascript
var config = {
  _index : 'pet'
};
```

Additionally, if working with multiple indexes, you may specify them as arrays:

```Javascript
var config = {
  _indices : ['pet', 'family'],
};
```

*Note:* When index or indices are supplied via operation options, those settings will take precedent over the base configuration for the library:

```Javascript

var
  elasticsearch = require('es'),
  config = {
    _index : 'kitteh'
  },
  es = elasticsearch.createClient(config);

es.indices.exist({ _index : 'canine' }, function (err, data) {
  // will result in a HEAD request to /canine instead of /kitteh
});
```

##### config.server

If omitted from configuration, the server settings default to the following:

```Javascript
var config = {
  // optional - when not supplied, defaults to the following:
  server : {
    host : 'localhost',
    port : 9200
  }
};
```

Anything specified within the server element of config is passed directly through to each HTTP/HTTPS request. You may configure additional options for connecting to Elasticsearch:

```Javascript
var config = {
  server : {
    agent : false,
    auth : 'user:pass',
    host : 'localhost',
    port : 9243,
    rejectUnauthorized : false,
    secure : true // toggles between https and http
  }
};
```

#### cluster support and failover

Elasticsearch is pretty much rad at clustering. If you want to specify multiple servers to failover to, you may do so by either supplying an array as the value for the property `host`, `hosts`, `hostname` or `hostnames`:

```Javascript
var elasticsearch = require('es');
var config = {
  _index : 'bawss',
  server : {
    hosts : ['es1.myhost.com', 'es2.myhost.com', 'es3.myhost.com']
    secure : true
  }
};

var es = elasticsearch(config);
```

If you run on different ports for each server, use the `hostnames` property:

```Javascript
var elasticsearch = require('es');
var config = {
  _index : 'bawss',
  server : {
    hostnames : ['localhost:9200', 'localhost:9201', 'localhost:9202']
  }
};

var es = elasticsearch(config);
```

#### operation timeout

The default timeout for any operation against Elasticsearch is set at 30 seconds. You can override this value by specifying a timeout property in the options for the operation:

```Javascript
var options = {
  timeout : 60000 // 60 seconds
};

es.bulk(options, commands, function (err, data) {
  // teh datas
});
```

#### request event

An event named `request` with a signature of `function (options) { }` is emitted for each API call.

```Javascript
var elasticsearch = require('es');

var config = {
  _index : 'bawss',
  server : {
    hosts : ['localhost:9200', 'localhost:9201', 'localhost:9202']
  }
};

var es = elasticsearch(config);

es.request.on('request', function (options) {
  console.log('request initiated');
  console.log(options);
});

es.count(function (err, results) {
  // event results in request options being logged to console...
});
```

#### options for any operation

For each ES operation, options may be specified as the first argument to the function. In most cases, these are entirely optional, but when supplied, the values specified will take precedent over the config values passed to the library constructor.
Additionally, if there are extra option keys supplied beyond what is required for the operation, they are mapped directly to the querystring.

```
var options = {
  _index : 'bawss',
  refresh : true
};

var doc = {
  field1 : 'test value'
};

es.index(options, doc, function (err, data) {
  // this will result in a POST with path /bawss/man?refresh=true
});
```

### Core

For more specifics and details regarding the core API for ElasticSearch, please refer to the documentation at <http://www.elasticsearch.org/guide/reference/api/>.

##### Bulk

*Please Note:* The default timeout is set at 30 seconds... if you are performing a large bulk insert you may need to increase this limit by specifying a higher value for `timeout` in the options parameter.

This method doesn't take into account the underlying config that was used when instantiating the client. It requires index to be specified via the commands array or via the options parameter. Conflict will occur if one specifies a different index in the options than what is specified via the commands parameter.

At a high level, when performing a bulk update, you must supply an array with an action object followed by the object that the action will use during execution. In the following example, the first item in the array specifies the action is `index` and the second item represents the data to index:

```Javascript
[
  { index : { _index : 'dieties' } },
  { name : 'hamish', breed : 'manx', color : 'tortoise' }
]
```

In this example, two `index` actions will be performed on the 'dieties' index in ElasticSearch:

```Javascript
[
  { index : { _index : 'dieties' } },
  { name : 'dugald', breed : 'siamese', color : 'white' },
  { index : { _index : 'dieties' } },
  { name : 'keelin', breed : 'domestic long-hair', color : 'russian blue' }
]
```

For more information regarding bulk, please see the ElasticSearch documentation at <http://www.elasticsearch.org/guide/reference/api/bulk/>

`es.bulk(options, commands, callback)`

```Javascript
var
  elasticsearch = require('es'),
  es = elasticsearch();

var commands = [
  { index : { _index : 'dieties' } },
  { name : 'hamish', breed : 'manx', color : 'tortoise' },
  { index : { _index : 'dieties' } },
  { name : 'dugald', breed : 'siamese', color : 'white' },
  { index : { _index : 'dieties' } },
  { name : 'keelin', breed : 'domestic long-hair', color : 'russian blue' }
];

es.bulk(options, commands, function (err, data) {
  // teh datas
});
```

##### Bulk Index

This is not a core action for ElasticSearch, but is a convenience method added to this ElasticSearch client to make bulk indexing more straight forward. Simply supply an array of documents you wish to bulk index in ElasticSearch and the method will take of the details for you.

`es.bulkIndex(options, documents, callback)`

```Javascript
var
  elasticsearch = require('es'),
  es = elasticsearch();

var documents = [
  { name : 'hamish', breed : 'manx', color : 'tortoise' },
  { name : 'dugald', breed : 'siamese', color : 'white' },
  { name : 'keelin', breed : 'domestic long-hair', color : 'russian blue' }
];

var options = {
  _index : 'dieties'
}

es.bulkIndex(options, documents, function (err, data) {
  // teh datas
});
```

##### Count

`es.count(options, query, callback)`

```Javascript
var
  elasticsearch = require('es');
  es = elasticsearch();

es.count(function (err, data) {
  // teh datas
});

// count docs in a specific index
var options = {
  _index : 'bawss'
}

es.count(options, function (err, data) {
  // counted... like a bawss
});

// count docs for a specific query
var query = {
  query : {
    term : {
      breed : 'manx'
    }
  }
};

es.count(options, query, function (err, data) {
  // teh count of teh manx kittehs
});
```

##### Delete

Requires `_index` be specified either via lib config (as shown below) or via options when calling the operation.

`es.delete(options, callback)`

```Javascript
var
  elasticsearch = require('es'),
  es = elasticsearch();

core.delete({ _id : 'mbQZc_XhQDWmNCQX5KwPeA' }, function (err, data) {
  // teh datas
});
```

##### Delete By Query

Requires `_index` be specified either via lib config (as shown below) or via options when calling the operation.

`es.deleteByQuery(options, query, callback)`

```Javascript
var
  elasticsearch = require('es'),
  es = elasticsearch({ _index : 'kitteh' });

var query = {
  query : {
    field : { breed : 'siamese' }
  }
};

es.deleteByQuery(query, function (err, data) {
  // teh datas
});
```

##### Exists

Requires `_index` be specified either via lib config or via options when calling the operation.

`es.exists(options, callback)`

```Javascript
var
  elasticsearch = require('es'),
  es = elasticsearch();

es.exists({ _index : 'kitteh' }, function (err, data) {
  // teh datas
});
```

##### Explain

Requires `_index` be specified either via lib config or via options when calling the operation.
Also requires `_id`, but this must be specified via options.

`es.explain(options, query, callback)`

##### Get

Requires `_index` be specified either via lib config or via options when calling the operation.
Also requires `_id`, but this must be specified via options.

`es.get(options, callback)`

##### Index

Requires `_index` be specified either via lib config or via options when calling the operation.

`es.index(options, doc, callback)`

##### More Like This

Requires `_index` be specified either via lib config or via options when calling the operation.
Also requires `_id`, but this must be specified via options.

`es.moreLikeThis(options, callback)`

##### Multi Get

If `_index` is supplied via options (or lib config), the will applied to the doc that is transmitted for the operation.

`es.multiGet(options, docs, callback)`

##### Multi Search

`es.multiSearch(options, queries, callback)`

##### Search

Requires `_index` be specified either via lib config or via options when calling the operation.

`es.search(options, query, callback)`

##### Scroll

Requires `scroll` be specified via options when calling the method. The following code snippet shows how to perform a search with a subsequent scroll.

```Javascript
var
  elasticsearch = require('es'),
  config = {
    _index : 'kittehs'
  },
  es = elasticsearch(config);

// first search
es.search({
    scroll : '10m'
  }, {
    query : {
      match_all : {}
    }
  }, function (err, data) {
    // next, perform the scroll with the _scroll_id value returned
    es.scroll({ scroll : '10m' }, data['_scroll_id'], callback);
  });
```

##### Suggest

`es.suggest(options, query, callback)`

##### Update

Requires `_index` be specified either via lib config or via options when calling the operation.
Also requires `_id`, but this must be specified via options.

`es.update(options, doc, callback)`

##### Validate

Requires `_index` be specified either via lib config or via options when calling the operation.

`es.validate(options, query, callback)`

### Indices

All operations here interact with the indices segment of the Elasticsearch API.

##### Alias

`es.indices.alias(options, data, callback)`

##### Aliases

Requires `alias`, but this must be specified via options.

`es.indices.aliases(options, callback)`

##### Analyze

`es.indices.analyze(options, data, callback)`

##### Clear Cache

`es.indices.clearCache(options, callback)`

##### Close Index

Requires `_index` be specified either via lib config or via options when calling the operation.

`es.indices.closeIndex(options, callback)`

##### Create Index

Requires `_index` be specified either via lib config or via options when calling the operation.

`es.indices.createIndex(options, data, callback)`

##### Create Template

Requires `name`, but this must be specified via options.

`es.indices.createTemplate(options, template, callback)`

##### Delete Alias

Requires `_index` and `_alias` be specified either via lib config or via options when calling the operation.

`es.indices.deleteAlias(opitons, callback)`

##### Delete Index

Requires `_index` be specified either via lib config or via options when calling the operation.

`es.indices.deleteIndex(options, callback)`

##### Delete Mapping

Requires `_index` be specified either via lib config or via options when calling the operation.

`es.indices.deleteMapping(options, callback)`

##### Delete Template

Requires `name`, but this must be specified via options.

`es.indices.deleteTemplate(options, callback)`

##### Exists

Requires `_index` be specified either via lib config or via options when calling the operation.

`es.indices.exists(options, callback)`

##### Flush

`es.indices.flush(options, callback)`

##### Mappings

`es.indices.mappings(options, callback)`

##### Open Index

Requires `_index` be specified either via lib config or via options when calling the operation.

`es.indices.openIndex(options, callback)`

##### Put Mapping

Requires `_index` be specified either via lib config or via options when calling the operation.

`es.indices.putMapping(options, mapping, callback)`

##### Refresh

`es.indices.refresh(options, callback)`

##### Segments

`es.indices.segments(options, callback)`

##### Settings

Requires `_index` be specified either via lib config or via options when calling the operation.

`es.indices.settings(options, callback)`

##### Snapshot

`es.indices.snapshot(options, callback)`

##### Stats

`es.indices.stats(options, callback)`

##### Templates

Requires `name`, but this must be specified via options.

`es.indices.templates(options, callback)`

##### Update Settings

`es.indices.updateSettings(options, settings, callback)`

### Cluster

All operations here interact with the Cluster portion of the Elasticsearch API.

##### Delete River

Requires `name`, but this must be specified via options.

`es.cluster.deleteRiver(options, callback)`

##### Field Stats

Requires `field` or `fields`, but this must be specified via options.

`es.cluster.fieldStats(options, callback)`

##### Health

`es.cluster.health(options, callback)`

##### Hot Threads

`es.cluster.hotThreads(options, callback)`

##### Nodes Info

`es.cluster.nodesInfo(options, callback)`

##### Nodes Stats

`es.cluster.nodesStats(options, callback)`

##### Put River

Requires `name`, but this must be specified via options.

`es.cluster.putRiver(options, meta, callback)`

##### Reroute

`es.cluster.reroute(options, commands, callback)`

##### Rivers

Requires `name`, but this must be specified via options.

`es.cluster.rivers(options, callback)`

##### Settings

`es.cluster.settings(options, callback)`

##### Shutdown

`es.cluster.shutdown(options, callback)`

##### State

`es.cluster.state(options, callback)`

##### Update Settings

`es.cluster.updateSettings(options, updates, callback)`

# Testing

Code coverage data generated from npm test is located in `./lib-cov` and is not included in the git repo.

```Javascript
npm install
npm test
```

To run code coverage and generate local report at `./reports/coverage.html`:

```Javascript
npm run-script coverage
```

# Requirements

* Node.js
* elasticsearch
* The need for search

# License

MIT


HISTORY.md
# v0.8.0 / 2021-06-23

* Adjustments to support Elasticsearch v7 and up (<https://www.elastic.co/guide/en/elasticsearch/reference/current/removal-of-types.html>)
  * Updated core.add (core.index): added `/_doc` to path and eliminated `_type` parameter requirement
  * Updated core.delete: added `/_doc` to path
  * Updated core.get: added `/_doc` to path and eliminated `_type` parameter requirement
  * Removed core.moreLikeThis: is now accomplished via search query (<https://www.elastic.co/guide/en/elasticsearch/reference/7.13/query-dsl-mlt-query.html>)
  * Updated core.multiGet: removed `_type` from path and eliminated `_type` parameter requirement
  * Updated core.update: removed `_type` from path and eliminated `_type` parameter requirement
  * Updated core.bulkIndex: removed `_type` from payload interpolation
  * Updated core.explain: removed `_type` from path and eliminated `_type` parameter requirement
  * Updated core.search: removed `_type` from path
  * Updated core.suggest: removed `_type` from path
  * Updated core.count: removed `_type` from path
  * Updated core.multiSearch: removed `_type` from path
  * Updated core.deleteByQuery: removed `_type` from path
  * Removed core.exists: fully deprecated in ES 7.x
  * Updated indices.putMapping: removed `/_mapping/_type` from resource and eliminated `_type` parameter requirement
  * Updated indices.stats: removed `_type` from path
  * Updated indices.deleteMapping: removed `_type` from path and elimited `_type` parameter requirement
  * Updated indices.exists: removed `_type` from path
  * Updated indices.mappings: removed `_type` from path

# v0.7.4 / 2021-06-22

* Updated dependencies

# v0.7.3 / 2019-03-19

* Modified `package.json` and moved to `prepare` instead of `prepublish`
* Updated `reqlib` dependency with tighter request body handling

# v0.7.2 / 2019-03-19

* Removed extraneous `console.log` statements in functional test

# v0.7.1 / 2019-03-05

* Adjusted `scroll` method to construct proper scroll payload when only the ID is specified

# v0.7.0 / 2019-02-20

* Replaced `thenify` and `bluebird` dependencies with `babel`
* Modified to support ElasticSearch 6.x.x
  * `suggest` method adjusted to use `_search` endpoint instead of deprecated `_suggest`
* Fixed issue where `cluster.hotThreads` callback returned result as `err`
* Adjusted failover behavior for `hosts` key, specifically to better align with the URL module in Node core
* Adjusted mappings in the functional tests according to changes in Elasticsearch 6.x

# v0.6.0 / 2017-10-02

* Modified to support ElasticSearch 5.5.x
  * `indices.createIndex` method adjusted to only call `put` (deprecated `post`)
  * `indices.optimize` method deprecated
  * `indices.status` method deprecated
  * `Warmers` deprecated
  * `Percolators` deprecated
  * `scan` search_type deprecated
  * `and` filter deprecate and replaced by `bool` query
  * unit and functional tests udpated

# v0.5.2 / 2016-10-05

* Added fix for scenario where `_index` and `_type` overrides would not always work (Issue #62)
* Adjusted functional tests to appropriately work with Elasticsearch versions greater than v2.0

# v0.5.1 / 2016-01-22

* Updated documentation for `count` method

# v0.5.0 / 2016-01-22

* Added support for promises while maintaining backwards compatibility in existing API (persuant to issue #58)
* Moved to eslint from jshint
* Factored mocha options into an options file
* Moved code coverage work to Travis-CI to eliminate developer side error when running `npm test`
* Moved to Istanbul from jscoverage and cleaned up NPM scripts used for testing

# v0.4.8 / 2015-10-15

* Fixed issue where `_source` parameter was not properly added to the URL when using `get`

# 0.4.7 / 2015-09-21

* Adjusted `es.indices.putMapping` to match [Elasticsearch documentation](<https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-put-mapping.html>)

# 0.4.6 / 2015-09-16

* Added support for `_create` as an option when calling `es.index` function

# 0.4.5 / 2015-08-06

* Added support for supplying additional parameters to `es.indices.putMapping` (i.e. `ignore_conflicts`)
* Minor updates to reflect additions and changes in versions of Elasticsearch post v1.3.x

# 0.4.4 / 2014-04-26

* Fix for issue #48 - querystring parameters as arrays were not properly formatted when sent

# 0.4.3 / 2014-04-22

* Fix for issue #47 - updating jscoverage dependency version

# 0.4.2 / 2014-04-22

* Fix for issue #46 - request defaults to port 80

# 0.4.1 / 2014-03-09

* Modifed to support ElasticSearch 1.0.x
  * `cluster.nodesInfo` method updated
  * `cluster.nodesStatus` method updated
  * `registerPercolator` method updated
  * `unregisterPercolator` method updated
  * unit and functional tests updated

# 0.3.16 / 2014-01-03

* Added support for `scroll()` method in core module

# 0.3.15 / 2013-12-21

* Fix to allow _id values of 0

# 0.3.14 / 2013-11-14

* Fix for update() method not allowing script field to be a blank value

# 0.3.13 / 2013-10-29

* Fix for EventEmitter listener Node warning

# 0.3.12 / 2013-10-02

* Switched module name to `es` from `elasticsearch`
* Fixed documentation spelling errors
* Added longer timeout for functional tests
* Added support for suggest API method

# 0.3.11 / 2013-09-19

* Rewrote mechanism for logging HTTP requests to use events instead
* Fixed bulk index method to handle passing in ids properly

# 0.3.10 / 2013-08-29

* Added ability to log HTTP requests

# 0.3.9 / 2013-08-27

* Expose underlying request API to clients for easier override
* Added functional test suite and configured in Travis-CI
* Fixed bug in multiGet method
* Fixed bug in deleteByQuery method

# 0.3.8 / 2013-08-15

* Changed underlying request and utils modules for better extensibility
* Added ability to override/extend core request module

# 0.3.7 / 2013-08-05

* Added bulkIndex method to core for convenience
* Enhanced documentation

# 0.3.6 / 2013-07-18


lib/request.js
var
	events = require('events'),
	http = require('http'),
	https = require('https'),
	qs = require('querystring'),

	DEFAULT_TIMEOUT = 30000,
	EVENT_REQUEST = 'request';


/*
	A closure that enables basic configuration of the request
	that will be applied for every operation of the lib.
*/
exports.initialize = function (settings, self) {
	'use strict';

	self = self || Object.create(events.EventEmitter.prototype);
	self.failover = {
		count : 0,
		index : 0,
		key : '',
		values : []
	};
	self.settings = settings || {};

	// enable events on request
	events.EventEmitter.call(self);

	/*
		If failover information is provided, store the options for use
		in the event of a connection failure.
	*/
	['hosts', 'hostnames'].forEach(function (hostsKey) {
		if (self.settings[hostsKey] && Array.isArray(self.settings[hostsKey])) {
			// remove the "s" from the key
			self.failover.key = hostsKey.slice(0, -1);
			self.failover.values = self.settings[hostsKey];

			// set the current default host/hostname
			self.settings[self.failover.key] = self.failover.values[self.failover.index];

			// check for port specification in host/hostname
			var portIndex = self.settings[self.failover.key].indexOf(':');
			if (portIndex > 0) {
				self.settings.port = self.settings[self.failover.key].substr(portIndex + 1);
				self.settings[self.failover.key] = self.settings[self.failover.key].substr(0, portIndex);
			}

			// clear the failover settings from the default settings
			delete self.settings[hostsKey];
		}
	});

	/*
		Actually executes a request given the supplied options,
		writing the specified data and returning the result to the
		supplied callback.

		In the event that an exception occurs on the request, the
		Error is captured and returned via the callback.
	*/
	function exec (options, data, callback) {
		options = self.getRequestOptions(options);
		data = data || '';
		if (typeof data !== 'string') {
			data = JSON.stringify(data);
		}
		if (!options.headers) {
			options.headers = {};
		}
		options.headers['Content-Length'] = Buffer.byteLength(data);

		// issue #23 - ability to log http(s) requests (emit an event update)
		if (self.emit) {
			self.emit(EVENT_REQUEST, options);
		}

		var req =
			(options.secure ? https : http).request(options, function (res) {
				var
					body = '',
					chunks = [],
					json = {},
					statusCode = res.statusCode;

				res.setEncoding('utf-8');

				res.on('data', function (chunk) {
					chunks.push(chunk);
				});

				res.once('end', function () {
					body = chunks.join('');

					if (statusCode >= 400) {
						var err = new Error(body);
						err.statusCode = statusCode;
						return callback(err);
					}

					if (!body && options.method === 'HEAD') {
						json = {
							statusCode : statusCode
						};
					} else {
						try {
							json = JSON.parse(body);
						} catch (err) {
							return callback(new Error(body, err));
						}
					}

					self.failover.count = 0;
					return callback(null, json);
				});
			});

		req.on('error', function (err) {
			var
				failover =
					self.failover.values.length > 1 &&
					err.code &&
					(err.code === 'ENOTFOUND' ||
					err.code === 'ECONNREFUSED' ||
					err.code === 'ECONNRESET'),
				hostOverridden =
					(options.hostname || options.host) !==
					(self.settings.hostname || self.settings.host);

			// if there is a connection error and the host/hostname settings
			// weren't overridden in the options, let's failover to the
			// next host/hostname if one is available
			if (failover && !hostOverridden) {
				self.failover.count++;
				self.failover.index =
					self.failover.index === self.failover.values.length - 1 ?
					0 :
					self.failover.index + 1;

				if (self.failover.count <= self.failover.values.length) {
					self.settings[self.failover.key] = self.failover.values[self.failover.index];
					options[self.failover.key] = self.settings[self.failover.key];

					return exec(options, data, callback);
				}
			}

			self.failover.count = 0;
			return callback(err, null);
		});

		// timeout the connection
		if (options.timeout) {
			req.setTimeout(options.timeout, function() {
				req.abort();
			});
		}

		if (data) {
			req.write(data);
		}

		req.end();

		return req;
	}

	/*
		Effectively merges request options with preconfigured
		information. Priority is given to the input options...

		This could get wonky if a client thinks to encapsulate
		settings for the server within a server sub-document of
		the options document.

		i.e.

		// this will override base config.host
		options.host = '127.0.0.1'

		// this will result in config.server being set... host
		// will not effectively be overriden for the request
		options.server.host = '127.0.0.1'
	*/
	self.getRequestOptions = function (options) {
		var returnOptions = {};

		Object.keys(self.settings).forEach(function (field) {
			returnOptions[field] = self.settings[field];
		});

		Object.keys(options).forEach(function (field) {
			returnOptions[field] = options[field];
		});

		// ensure default timeout is applied if one is not supplied
		if (typeof returnOptions.timeout === 'undefined') {
			returnOptions.timeout = DEFAULT_TIMEOUT;
		}

		// create `path` from pathname and query.
		returnOptions.path = returnOptions.pathname;
		if (returnOptions.query && Object.keys(returnOptions.query).length) {
			// fix for #48 (arrays in querystring are properly serialized)
			Object.keys(returnOptions.query).forEach(function (param) {
				if (Array.isArray(returnOptions.query[param]) && returnOptions.query[param].length > 1) {
					returnOptions.query[param] = returnOptions.query[param].join(',');
				}
			});

			returnOptions.path += '?' + qs.stringify(returnOptions.query);
		}

		return returnOptions;
	};

	/*
		Issues a DELETE request with data (if supplied) to the server

		Disclaimer: It's not a common practice to pass POST
		data via the DELETE method. This, however, is how the
		ES API operates:

		http://www.elasticsearch.org/guide/reference/api/delete-by-query/
	*/
	self.delete = function (options, data, callback) {
		if (!callback && typeof data === 'function') {
			callback = data;
			data = null;
		}

		options.method = 'DELETE';
		return exec(options, data, callback);
	};

	/*
		Custom request with data to the server
	*/
	self.exec = function (options, data, callback) {
		if (!callback && typeof data === 'function') {
			callback = data;
			data = null;
		}

		return exec(options, data, callback);
	};

	/*
		Issues a GET request with data (if supplied) to the server

		Disclaimer: It's not a common practice to pass POST
		data via the GET method. This, however, is how the
		ES API operates:

		http://www.elasticsearch.org/guide/reference/api/search/
		http://www.elasticsearch.org/guide/reference/api/multi-get/
		http://www.elasticsearch.org/guide/reference/api/multi-search/
		http://www.elasticsearch.org/guide/reference/api/percolate/
		http://www.elasticsearch.org/guide/reference/api/validate/
	*/
	self.get = function (options, data, callback) {
		if (!callback && typeof data === 'function') {
			callback = data;
			data = null;
		}

		options.method = 'GET';
		return exec(options, data, callback);
	};

	/*
		Issues a HEAD request with data (if supplied) to the server

		Disclaimer: It's not a common practice to pass POST
		data via the HEAD method.

		It's in here, however, because ES does this for other
		methods (i.e. GET and DELETE)... there is no immediate
		need for this.
	*/
	self.head = function (options, data, callback) {
		if (!callback && typeof data === 'function') {
			callback = data;
			data = null;
		}

		options.method = 'HEAD';
		return exec(options, data, callback);
	};

	/*
		Issues a POST request with data (if supplied) to the server
	*/
	self.post = function (options, data, callback) {
		if (!callback && typeof data === 'function') {
			callback = data;
			data = null;
		}

		options.method = 'POST';
		return exec(options, data, callback);
	};

	/*
		Issues a PUT request with data (if supplied) to the server
	*/
	self.put = function (options, data, callback) {
		if (!callback && typeof data === 'function') {
			callback = data;
			data = null;
		}

		options.method = 'PUT';
		return exec(options, data, callback);
	};

	return self;
};
